### PR TITLE
Fixed #25176 -- Prevented a failure in TestCase.setUpTestData() from leaking the transaction.

### PR DESCRIPTION
**Pull request by adamchainz**
[Ticket 25176](https://code.djangoproject.com/ticket/25176#ticket)

### CODE REVIEW

File: django/test/testcases.py [hunk 1]
```diff
@@ -999,7 +999,11 @@ def setUpClass(cls):
                     except Exception:
                         cls._rollback_atomics(cls.cls_atomics)
```
Comment (aaugustin) [inline django/test/testcases.py]: You should probably inject `*sys.exc_info()` here as well.

Comment (adamchainz) [inline django/test/testcases.py]: Well spotted, thanks :rainbow:

**Comment by aaugustin** [comment]
I looked more in detail it this patch today. (Yesterday's comment was just about an inconsistency I spotted while skimming the patch.)

Catching exceptions in `setUpTestData` and rolling back on exceptions makes sense.

Changing the signature of `_rollback_atomics` isn't needed, as far as I can tell. `_rollback_atomics` sets the rollback flag, which is enough to guarantee a rollback regardless of the exception state. What was the reason for making this change?

**Comment by adamchainz** [comment]
I was just being thorough by feeding the exception info to `atomic` - I figured that its behaviour might depend on whether we are exiting an exception. I skimmed it and saw `exc_type` is inspected, but I didn't exactly what was going on.

**Comment by aaugustin** [comment]
The condition tested is `exc_type is None and not connection.needs_rollback`.

In the scenario we're discussing here, `connection.needs_rollback` is `True` (that's the rollback flag I was referring to in my previous comment) so the value of exc_type doesn't matter.

Considering the guarantees made by the current transaction management public APIs, you don't need to pass `sys.exc_info()` to `_rollback_atomics`.

**Comment by adamchainz** [comment]
Okay cool, thanks for clearing that up for me. Will rewrite and re-attempt to find a test, as tim noted on the ticket, shortly.

**Comment by timgraham** [comment]
Here's what I had in mind for a test. It works as a regression test as far as I could tell.
```diff
diff --git a/tests/test_utils/tests.py b/tests/test_utils/tests.py
index 7056ea3..0844288 100644
--- a/tests/test_utils/tests.py
+++ b/tests/test_utils/tests.py
@@ -916,6 +916,39 @@ class OverrideSettingsTests(SimpleTestCase):
             self.assertIn(expected_location, finder.locations)
 
 
+class TestBadSetupTestData(TestCase):
+    """
+    An exception in setUpTestData() shouldn't leak a transaction which would
+    cascade across the rest of the test suite.
+    """
+    class MyException(Exception):
+        pass
+
+    @classmethod
+    def setUpClass(cls):
+        try:
+            super(TestBadSetupTestData, cls).setUpClass()
+        except cls.MyException:
+            cls._in_atomic_block = connection.in_atomic_block
+
+    @classmethod
+    def tearDownClass(Cls):
+        # override to avoid a second cls._rollback_atomics() which would fail.
+        # Normal setUpClass() methods won't have exception handling so this
+        # method wouldn't typically be run.
+        pass
+
+    @classmethod
+    def setUpTestData(cls):
+        # Simulate a broken setUpTestData() method.
+        raise cls.MyException()
+
+    def test_failure_in_setUpTestData_should_rollback_transaction(self):
+        # setUpTestData() should call _rollback_atomics() so that the
+        # transaction doesn't leak.
+        self.assertFalse(self._in_atomic_block)
+
+
 class DisallowedDatabaseQueriesTests(SimpleTestCase):
     def test_disallowed_database_queries(self):
         expected_message = (
```

**Comment by timgraham** [comment]
We could backport this to 1.8, so please add 1.8.4 release notes when updating. Please mark ticket "Ready for checkin" when done, thanks.

**Comment by adamchainz** [comment]
I was a bit dubious about that test, since it bypasses a lot of `unittest`'s assumptions. For example, `setUp` and `tearDown` will still by run. Even if this works today, future test harness changes, in or outside of django, might leave this with a potentially mysterious error. Anyway, I guess there's relatively little harm in adding it since it should get noticed.

**Comment by timgraham** [comment]
merged in 0abb06930fc0686cb35079934e5bb40df66f5691, thanks!

**Comment by adamchainz** [comment]
Cool. Sorry I missed the release notes, hope I haven't made anything difficult.

**Pull request closed without merging**